### PR TITLE
Cache eviction for single execution in datacatalog and flyteadmin

### PR DESCRIPTION
**Pull request by MorpheusXAUT**
# TL;DR
This PR prepares datacatalog and flyteadmin endpoints for cache eviction changes, overwriting existing artifacts for a single execution.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The new `UpdateArtifact` RPC endpoint for the datacatalog service allows for an existing `Artifact` to be updated. At the moment, this update can only contain a (complete) replacement of the associated `ArtifactData`.
Artifacts can be identified using their ID or tag.

The new `skip_cache` flag in flyteadmin's `ExecutionSpec`, `LaunchPlanSpec` and `WorkflowExecutionConfig` are used to implement the [cache eviction override for a single execution](https://hackmd.io/qOztkaj4Rb6ypodvGEowAg?view#Cache-eviction-override-for-a-single-execution).

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2867

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by MorpheusXAUT** [comment]
Pushed additional IDL changes for flyteadmin, title/initial description have been adapted accordingly.

**Comment by katrogan** [comment]
hey @MorpheusXAUT this change looks good but I'm curious if you're planning on implementing the back-end changes to also include cache eviction and skipping the cache in the same PR per component. maybe it would make more sense to split these changes out individually (eviction & skipping)

also, before merging this would it be possible to have the other back-end PRs prepared and tested end to end to reduce churn in updating flyteidl (especially when it comes to potentially having to deprecate proto fields if the message structure ends up changing over the course of your testing)

**Comment by MorpheusXAUT** [comment]
Hi @katrogan , thanks for taking a look!

> hey @MorpheusXAUT this change looks good but I'm curious if you're planning on implementing the back-end changes to also include cache eviction and skipping the cache in the same PR per component. maybe it would make more sense to split these changes out individually (eviction & skipping)

Yeah, I was planning on making a PR about skipping cache only (and everything that's required for that) and open another one for the second part of the implementing (evicting cached data).

> also, before merging this would it be possible to have the other back-end PRs prepared and tested end to end to reduce churn in updating flyteidl (especially when it comes to potentially having to deprecate proto fields if the message structure ends up changing over the course of your testing)

Sure thing! I'm mostly done with the other PRs, currently finishing up flytectl and flyteconsole client implementations. The rest of the backend implementation is done, I'll open PRs for those tomorrow (using `replace` to our forks in `go.mod` to it'll build until this PR is merged).

I'll mention this PR in the others as well so you'll have a quick overview of all changes once they're open.

**Comment by MorpheusXAUT** [comment]
@katrogan All other PRs have been submitted, created as drafts for now since they depend on this PR/each other.

Pushed some minor changes to protos as well today as they were required for the `flytectl` changes.

**Review by hamersaw — COMMENTED** [review]
Left a few comments on the [flytepropeller PR](https://github.com/flyteorg/flytepropeller/pull/489) - Basically, I'm concerned about using `skipCache` for the naming of this feature. Really we are skipping the cache lookup but overwriting the existing cache values - would `overwriteCache` be a better indicator of the functionality? I think leaving `CACHE_SKIPPED` status during the lookup is correct and the rest of this PR lgtm. Open for discussion on this.

**Review by hamersaw — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
